src/main/java/org/jabref/gui/JabRefFrame.java

```diff
@@ -492,12 +493,42 @@ public boolean quit() {
         WaitForSaveFinishedDialog waitForSaveFinishedDialog = new WaitForSaveFinishedDialog(dialogService);
         waitForSaveFinishedDialog.showAndWait(getLibraryTabs());
 
+        // We call saveWindow state here again because under Mac the windowClose listener on the stage isn't triggered when using cmd + q
+        saveWindowState(mainStage);
         // Good bye!
         tearDownJabRef(filenames);
         Platform.exit();
         return true;
     }
 
+    private void saveWindowState(Stage mainStage) {
+        GuiPreferences preferences = prefs.getGuiPreferences();
+        preferences.setPositionX(mainStage.getX());
+        preferences.setPositionY(mainStage.getY());
+        preferences.setSizeX(mainStage.getWidth());
+        preferences.setSizeY(mainStage.getHeight());
+        preferences.setWindowMaximised(mainStage.isMaximized());
+        preferences.setWindowFullScreen(mainStage.isFullScreen());
+        debugLogWindowState(mainStage);
+    }
```

Comment:
Hm, this is not part of the frame, but of the stage. I was hoping I could untangle all that GUI-CLI-opendatabase stuff on jabcon. Right now this is just a code duplication of JabRefGUI#saveWindowState. But maybe you can make this method in JabRefFrame public and remove the one in JabRefGui?

Reply:
Good idea. The key problem is that cmd + Q (like alt + f4 ) does not trigger the onCloseRequest listener of the stage

Reply:
Yes, i did a little bit of reading on this one. I believe the main problem is that the whole startup and shutdown process has to be rewritten.
JabRefGui and MainApplication should probably be merged, the creation of ExecutorService be moved to JabRefGui so it can be cleanly be shutdown, so we can solve the problem with the runaway process when all windows are closed (see also `Platform.setImplicitExit(true);` ), the different openDatabase methods be unified ...

Reply:
We had this before, but then the decision was made to avoid init of javafx directly and thus we have this Gui and CLI etc.

Reply:
> We had this before, but then the decision was made to avoid init of javafx directly and thus we have this Gui and CLI etc.

MainApplication is part of the GUI component. The Launcher class is the super class deciding on opening GUI or cli.

About MacOS stuff: parts of the codebase and parts of the architecture of jabref are 20 years old. It was necessary at one point, as it has fixed a bug once. Would probably have to be tested if this is still necessary.